src/parser.rs

```diff
@@ -407,6 +411,243 @@ pub fn parse_gcov(gcov_path: &Path) -> Result<Vec<(String,CovResult)>, ParserErr
+                    "class" => {
+                        // Find class name
+                        let fq_class = get_xml_attribute(&attributes, "name").unwrap(); // org/example/Person$Age
+                        let class = fq_class.split('/').last().unwrap();                // Person$Age
+                        let top_class = class.clone().split("$").nth(0).unwrap();       // Person
+
+                        // Process all <method /> and <counter /> for this class
+                        let functions = parse_java_class(parser, class).unwrap();
+
+                        match results_hash.entry(top_class.to_string()) {
+                            hash_map::Entry::Occupied(obj) => {
+                                obj.into_mut().functions.extend(functions);
+                            },
+                            hash_map::Entry::Vacant(v) => {
+                                v.insert(CovResult {
+                                    functions,
+                                    lines: BTreeMap::new(),
+                                    branches: BTreeMap::new(),
+                                });
+                            },
+                        };
+                    },
+                    "sourcefile" => {
+                        // Find source file name
+                        let sourcefile = get_xml_attribute(&attributes, "name").unwrap(); // Person.java
+                        let class = sourcefile.trim_right_matches(".java"); // Person
+
+                        // Process all <line /> and <counter /> for this class
+                        let (lines, branches) = parse_java_sourcefile(parser).unwrap();
+
+                        // Save the result
+                        match results_hash.entry(class.to_string()) {
+                            hash_map::Entry::Occupied(obj) => {
+                                let obj = obj.into_mut();
+                                obj.lines = lines;
+                                obj.branches = branches;
+                            },
+                            hash_map::Entry::Vacant(v) => {
+                                v.insert(CovResult {
+                                    functions: HashMap::new(),
+                                    lines,
+                                    branches,
+                                });
+                            },
+                        };
+                    },
+                    &_ => {},
+                }
+            },
+            Ok(XmlEvent::EndElement{name}) => {
+                match name.local_name.as_str() {
+                    "package" => break,
+                    &_ => {}
+                }
+            },
+            Err(e) => {
+                return Err(ParserError::Parse(e.to_string()))
+            },
+            _ => {}
+        }
+    }
+
+    Ok(
+        results_hash.into_iter().filter(
+            |(_, result)| result.lines.len() > 0 || result.branches.len() > 0
```

Comment:
Are there classes with no line information at all?

Reply:
Didn't find any classes with no line information. Instead, what I found were a few classes that were not the top class or inner class in their file. Those kind of classes are also removed by this filter.

This means we will need to figure out the mapping between java files and classes contained in them.

Reply:
Oh that's a pity. An example?

Reply:
These are the files that contain multiple top-level java classes:

- [mobile/android/base/java/org/mozilla/gecko/RemotePresentationService.java](https://hg.mozilla.org/mozilla-central/file/0e0813df0f65/mobile/android/base/java/org/mozilla/gecko/RemotePresentationService.java)
        - org/mozilla/gecko/VirtualPresentation
- [mobile/android/base/java/org/mozilla/gecko/mdns/MulticastDNSManager.java](https://hg.mozilla.org/mozilla-central/file/0e0813df0f65/mobile/android/base/java/org/mozilla/gecko/mdns/MulticastDNSManager.java)
        - org/mozilla/gecko/mdns/DiscoveryListener
        - org/mozilla/gecko/mdns/DummyMulticastDNSManager
        - org/mozilla/gecko/mdns/RegistrationListener
        - org/mozilla/gecko/mdns/NsdMulticastDNSManager
        - org/mozilla/gecko/mdns/ResolveListener
        - org/mozilla/gecko/mdns/MulticastDNSEventManager